fabtests/multinode/src/core.c

```diff
@@ -247,6 +251,11 @@ int multi_msg_send()
 
 		offset = state.sends_posted % opts.window_size;
 		assert(tx_ctx_arr[offset].state == OP_DONE);
+		if (ft_check_opts(FT_OPT_PERF)) {
+			timer_index = state.iter * pm_job.num_ranks +
+				(state.sends_posted % pm_job.num_ranks);
+			multi_timer_start(timers[timer_index]);
```

Comment:
I have no idea what timer_index represents or if this is correct.  I *think* timers would be better defined as some sort of multi-dimensional array, but I can't tell how the array is structured.  For example, is it this (conceptually)?
```
struct timestamp {
    uint64_t start_time;
    uint64_t stop_time;
};

struct timestamp times[RANKS][ITER];
```
in which case, this could be:
```
#define START_INDEX 0
#define STOP_INDEX 1
uint64_t times[RANKS][ITER][2];
```

Reply:
That is basically how it is structured currently. I set it up this way to make it easier to move all the timers through a sockets call. The 2d array is much easier to read, and this is a lot more readable. 

Does the concern about sending the timers even matter?

Reply:
I would wrap the timestamps in some sort of 2D array abstraction.  Something sort of like this:
```
ts_alloc(max_ranks, max_entries)
{
    ts_array = calloc(max_ranks * max_entries, sizeof(*ts_array));
    ts_array->max_ranks = max_ranks;
    ts_array->max_entries = max_entries;
}

ts_entry(ts_array, rank, index)
{
    assert(rank < ts_array->max_rank && index < ts_array->max_entries);
    return ts_array + rank * ts_array->max_entries + index;
}
```
then you could do:
```
ts_entry(ts_array, cur_rank, iter)->start = ft_gettime_ns();
...
ts_entry(ts_array, cur_rank, iter)->stop = ft_gettime_ns();
```
This makes it clearer that we're dealing with an array of timestamps per rank, and removes the questionable use of the modulo operator in finding the entry.

Reply:
This makes sense, I will reimplement this way

Reply:
As I'm implementing this, I'm not actually sure that this saves any sort of index magic. In the current implementation the timers are ordered the way they are used, num_ranks per iteration (at maximum, some patterns use fewer). A 2d array implementation does look a tad cleaner, but the modulo just ends up inside of the index directly.

So to start a given timer it would look like: `timer_start(timers[iter][state.sends_posted % pm_job.num_ranks])`, which is slightly cleaner given the lack of multiplication on the first index.

Moving away from all of these structs does make sense in that it saves a lot of malloc calls, but I don't know if it does the readability improvement you're looking for.

Reply:
What I'm looking for is an easy and direct mapping for the array entry that a given rank uses for a specific send.  It's the modulo that's confusing things.  See the ts_entry() function I listed above.  This assumes that we can calculate the total number of timestamps that we need up front in order to do one allocation.